File: wiki_java/src/main/java/com/github/klee0kai/wiki/wrapping/CustomLazy.java
package com.github.klee0kai.wiki.wrapping;

import com.github.klee0kai.stone.wrappers.Ref;

import java.lang.ref.WeakReference;

public class CustomLazy<T> {

    private T value = null;
    private final WeakReference<T> call;

    public CustomLazy(Ref<T> call) {
        this.call = new WeakReference(call.get());
    }

    public T getValue() {
        if (value != null)
            return value;
        return value = call.get();
    }

}

File: wiki_kotlin/src/main/kotlin/com/github/klee0kai/wiki/wrapping/CustomLazy.kt
package com.github.klee0kai.wiki.wrapping

import com.github.klee0kai.stone.weakref.Ref
import java.lang.ref.WeakReference

class CustomLazy<T>(call: Ref<T>) {

    private val weakRef: WeakReference<T> = WeakReference<T>(call.get())

    val value: T? get() = weakRef.get()

}